#1063 eo-maven-plugin:clean

l3r8yJ opened the pull request:
@Graur take a look, please

#1063

Graur commented on eo-maven-plugin/src/main/java/org/eolang/maven/CleanMojo.java:
```diff
@@ -0,0 +1,81 @@
+            Logger.info(this, "Target directory not exist.");
+            return;
+        }
+        try (Stream<Path> paths = Files.walk(this.targetDir.toPath())) {
```
@l3r8yJ I believe it would be better to use `org.eolang.maven.Walk` for this purpose. Check how it was done in `UnspileMojo`, please

l3r8yJ replied:
@Graur done

l3r8yJ replied:
Am I using the `Walk` class correctly?

```
new Walk(this.targetDir.toPath())
    .stream()
    .sorted(Comparator.reverseOrder())
    .map(Path::toFile)
    .forEach(CleanMojo::purge);
```

l3r8yJ replied:
because when i'm run the test getting a message that the file is not deleted and it's fails

Graur commented on eo-maven-plugin/src/main/java/org/eolang/maven/CleanMojo.java:
```diff
@@ -0,0 +1,81 @@
+        } catch (final IOException ex) {
+            Logger.error(
+                CleanMojo.class,
+                ex.getMessage()
```
@l3r8yJ Maybe we can do an error message more meaningful?

l3r8yJ replied:
Done

Graur commented on eo-maven-plugin/src/main/java/org/eolang/maven/CleanMojo.java:
```diff
@@ -0,0 +1,81 @@
+    @Override
+    final void exec() throws IOException {
+        if (!this.targetDir.exists()) {
+            Logger.info(this, "Target directory not exist.");
```
@l3r8yJ Can we add the name of this directory?

l3r8yJ replied:
done

Graur commented on eo-maven-plugin/src/test/java/org/eolang/maven/CleanMojoTest.java:
```diff
@@ -0,0 +1,58 @@
+            Matchers.is(true)
+        );
+    }
+}
```
@l3r8yJ Can you add test with full compiling lifecycle?

l3r8yJ replied:
done

Graur reviewed (changes requested):
@l3r8yJ Good one, just a few comments below:

l3r8yJ commented:
Added some logging:

Case with static method
```
try(Stream<Path> paths = Files.walk(this.targetDir.toPath())) {
    paths.sorted(Comparator.reverseOrder())
        .map(Path::toFile)
        .forEach(dir -> {
            purge(dir);
            Logger.info(
                this,
                "purged %s",
                dir.toString()
            );
        });
}
```
Logs:
> ```
> [INFO] org.eolang.maven.CleanMojo: purged /var/folders/3f/gxppnjks1f1crcj40mh04c700000gn/T/junit11832602380216143546/target/some16270794897569917092.eo
> [INFO] org.eolang.maven.CleanMojo: purged /var/folders/3f/gxppnjks1f1crcj40mh04c700000gn/T/junit11832602380216143546/target/child/child.eo
> [INFO] org.eolang.maven.CleanMojo: purged /var/folders/3f/gxppnjks1f1crcj40mh04c700000gn/T/junit11832602380216143546/target/child
> [INFO] org.eolang.maven.CleanMojo: purged /var/folders/3f/gxppnjks1f1crcj40mh04c700000gn/T/junit11832602380216143546/target
> [INFO] org.eolang.maven.CleanMojo: Deleted target: /var/folders/3f/gxppnjks1f1crcj40mh04c700000gn/T/junit11832602380216143546/target
> ```

Case with `Walk` class
```
new Walk(this.targetDir.toPath())
            .stream()
            .map(Path::toFile)
            .forEach(
                dir ->{
                    purge(dir);
                    Logger.info(
                        this,
                        "purged %s",
                        dir.toString()
                    );
                }
            );
```
[INFO] org.eolang.maven.CleanMojo: purged /var/folders/3f/gxppnjks1f1crcj40mh04c700000gn/T/junit5920592594147381947/target/some17529429556376472717.eo
[INFO] org.eolang.maven.CleanMojo: Deleted target: /var/folders/3f/gxppnjks1f1crcj40mh04c700000gn/T/junit5920592594147381947/target

for some reason it's not iterates trough all dirs, @Graur do you have any idea what's causing this?

l3r8yJ commented:
Added some logging:

Case with static method
```
try(Stream<Path> paths = Files.walk(this.targetDir.toPath())) {
    paths.sorted(Comparator.reverseOrder())
        .map(Path::toFile)
        .forEach(dir -> {
            purge(dir);
            Logger.info(
                this,
                "purged %s",
                dir.toString()
            );
        });
}
```
Logs:
> ```
> [INFO] org.eolang.maven.CleanMojo: purged /var/folders/3f/gxppnjks1f1crcj40mh04c700000gn/T/junit11832602380216143546/target/some16270794897569917092.eo
> [INFO] org.eolang.maven.CleanMojo: purged /var/folders/3f/gxppnjks1f1crcj40mh04c700000gn/T/junit11832602380216143546/target/child/child.eo
> [INFO] org.eolang.maven.CleanMojo: purged /var/folders/3f/gxppnjks1f1crcj40mh04c700000gn/T/junit11832602380216143546/target/child
> [INFO] org.eolang.maven.CleanMojo: purged /var/folders/3f/gxppnjks1f1crcj40mh04c700000gn/T/junit11832602380216143546/target
> [INFO] org.eolang.maven.CleanMojo: Deleted target: /var/folders/3f/gxppnjks1f1crcj40mh04c700000gn/T/junit11832602380216143546/target
> ```

Case with `Walk` class
```
new Walk(this.targetDir.toPath())
    .stream()
    .sorted(Comparator.reverseOrder())
    .map(Path::toFile)
    .forEach(
        dir ->{
            purge(dir);
            Logger.info(
                this,
                "purged %s",
                dir.toString()
            );
        }
    );
```

Logs:
```
> [INFO] org.eolang.maven.CleanMojo: purged /var/folders/3f/gxppnjks1f1crcj40mh04c700000gn/T/junit5920592594147381947/target/some17529429556376472717.eo
> [INFO] org.eolang.maven.CleanMojo: Deleted target: /var/folders/3f/gxppnjks1f1crcj40mh04c700000gn/T/junit5920592594147381947/target
```

for some reason it's not iterates trough all dirs, @Graur do you have any idea what's causing this?

l3r8yJ commented:
@Graur i'm found way to fix this, created new method in `Walk` class, and call it `reversed()`, what do you think? 

```
    /**
     * The reversed walk.
     * @return New Walk reversed
     * @throws IOException If fails
     */
    Walk reversed() throws IOException {
        return new Walk(
            this.home,
            Files.walk(this.home)
                .sorted(Comparator.reverseOrder())
                .collect(Collectors.toList())
        );
    }
```

l3r8yJ commented:
@Graur take a look, please.
On my local machine the build succeeds, but on CI `Save` does not exist? 
```
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary for eo 1.0-SNAPSHOT:
[INFO] 
[INFO] eo ................................................. SUCCESS [  1.309 s]
[INFO] eo-parser .......................................... SUCCESS [ 10.673 s]
[INFO] eo-maven-plugin .................................... SUCCESS [ 50.830 s]
[INFO] eo-runtime ......................................... SUCCESS [ 37.005 s]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  01:40 min
[INFO] Finished at: 2022-09-21T15:11:54+03:00
[INFO] ------------------------------------------------------------------------
```

Graur commented:
@l3r8yJ Please, do merge from upstream/master branch to your local branch and try again

l3r8yJ commented:
@Graur we passed, take a look, please)

Graur commented on eo-maven-plugin/src/main/java/org/eolang/maven/CleanMojo.java:
```diff
@@ -0,0 +1,101 @@
+        try {
+            Files.deleteIfExists(file.toPath());
+        } catch (final IOException ex) {
+            Logger.error(
```
@l3r8yJ This is called exception swallowing, we are trying to avoid it. Please, read this: https://www.yegor256.com/2015/07/09/catch-if-cant-otherwise.html

l3r8yJ replied:
done

Graur commented on eo-maven-plugin/src/main/java/org/eolang/maven/CleanMojo.java:
```diff
@@ -0,0 +1,101 @@
+     * @param dir The directory
+     */
+    private void log(final String msg, final File dir) {
+        Logger.info(
```
@l3r8yJ I don't this this log level is a good choice, maybe we can change it to `debug` ?

Graur reviewed (changes requested):
@l3r8yJ Good one, just a few comments below:

Graur reviewed (approved):
@yegor256 please, take a look

yegor256 commented on eo-maven-plugin/src/main/java/org/eolang/maven/CleanMojo.java:
```diff
@@ -0,0 +1,101 @@
+     * @param msg Message for logging
+     * @param dir The directory
+     */
+    private void log(final String msg, final File dir) {
```
@l3r8yJ why do you need this method? can't you call `Logger.debug` where it's needed?

yegor256 commented on eo-maven-plugin/src/main/java/org/eolang/maven/CleanMojo.java:
```diff
@@ -0,0 +1,101 @@
+            this.log("Directory %s isn't exist.", targetDir);
+            return;
+        }
+        new Walk(this.targetDir.toPath())
```
@l3r8yJ can't we use some existing method for directory recursive deletion, from JDK?

l3r8yJ replied:
@yegor256 should i remove the `reversed` method from `Walk`?

yegor256 commented:
@l3r8yJ I'm sorry about the delay. Please, check my comments above.

l3r8yJ commented:
@yegor256 sure

l3r8yJ commented:
@yegor256 have a look, please

yegor256 commented:
@l3r8yJ @Graur not yet perfect, see above

l3r8yJ commented:
@Graur have a look, please

Graur reviewed (approved):
@l3r8yJ Much better, thanks!

l3r8yJ commented:
@yegor256 have a look, please

yegor256 commented:
@rultor merge

rultor commented:
> @rultor merge

@yegor256 OK, I'll try to merge now. You can check the progress of the merge [here](https://www.rultor.com/t/27560-1284257070)

Graur commented:
@l3r8yJ Thanks!
